initialization/initializer.go

```diff
@@ -0,0 +1,547 @@
+			}
+		}()
+	}
+
+	outputs := make([]*MTreeOutput, numFiles)
+	for i := 0; i < numFiles; i++ {
+		select {
+		case res := <-resultsChan:
+			outputs[res.Index] = res.MTreeOutput
+		case err := <-errChan:
+			return nil, err
+		}
+	}
+
+	return outputs, nil
+}
+
+func (init *Initializer) initFile(fileIndex int, labelGroupsPerFile uint64, infileParallelism int) (*MTreeOutput, error) {
+	// Initialize the labels file writer.
+	labelsWriter, err := persistence.NewLabelsWriter(init.cfg.DataDir, init.id, fileIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize the labels merkle tree with the execution-phase zero challenge.
+	cacheWriter := cache.NewWriter(cache.MinHeightPolicy(init.cfg.LowestLayerToCacheDuringProofGeneration), cache.MakeSliceReadWriterFactory())
+	tree, err := merkle.NewTreeBuilder().
+		WithHashFunc(shared.ZeroChallenge.GetSha256Parent).
+		WithCacheWriter(cacheWriter).
+		Build()
+	if err != nil {
+		return nil, err
+	}
+
+	// Potentially perform recovery procedure.
+	// if file already exists, read it from start, and re-construct the merkle tree state.
+	// If file initialization was complete, return the merkle tree output.
+	// Otherwise continue to initialize from latest position.
+
+	labelsReader, err := labelsWriter.GetReader()
+	if err != nil {
+		return nil, err
+	}
+
+	existingWidth, err := labelsReader.Width()
+	if err != nil {
+		return nil, err
+	}
+
+	if existingWidth > 0 {
+		if existingWidth > labelGroupsPerFile {
+			return nil, ErrStateInconsistent
+		}
+
+		init.logger.Info("initialization recovery: starting file %v, position: %v, missing: %v", fileIndex, existingWidth, labelGroupsPerFile-existingWidth)
+
+		for {
+			lg, err := labelsReader.ReadNext()
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				return nil, err
+			}
+
+			err = tree.AddLeaf(lg)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		if existingWidth == labelGroupsPerFile {
+			cacheWriter.SetLayer(0, labelsReader)
+			cacheReader, err := cacheWriter.GetReader()
+			if err != nil {
+				return nil, err
+			}
+
+			return &MTreeOutput{Reader: cacheReader, Root: tree.Root()}, nil
+		}
+	}
+
+	init.logger.Info("initialization: start writing file %v, parallelism degree: %v",
+		fileIndex, infileParallelism)
+
+	numWorkers := infileParallelism
+	workersChans := make([]chan [][]byte, numWorkers)
+	errChan := make(chan error, 0)
+	finishedChan := make(chan struct{}, 0)
+	batchSize := 100
+	chanBufferSize := 100
+
+	// CPU workers.
+	fileOffset := uint64(fileIndex) * labelGroupsPerFile
+	for i := 0; i < numWorkers; i++ {
+		i := i
+		workersChans[i] = make(chan [][]byte, chanBufferSize)
+		workerOffset := i
+		go func() {
+			// Calculate labels in groups and write them to channel in batches.
+			position := uint64(workerOffset*batchSize) + existingWidth
+			batch := make([][]byte, batchSize)
+			batchPosition := uint64(0)
+
+			for batchPosition+position < labelGroupsPerFile {
```

Comment:
I think it would really help readability if this section had some clarifying comments. Wasn't sure where to place this comment - it's relevant to the whole section incl. the external `for` loop.